### ISSUE TITLE
Use directories instead of dashes

### ISSUE THREAD

**Issue by evangelion1204** [lorenwest/node-config#294]
Hi,

I just wondered if it would make sense to enhance the lib to support directories if an `index.js` is in there (a simple `require` in the end) I already saw #234 but I think thats a bit different. 

Btw I wouldn't mind to implement it on my own.
```
default.EXT => require('./default')
default-{instance}.EXT => require('./default/{instance}')
{deployment}.EXT => require('./{deployment}')
{deployment}-{instance}.EXT => require('./{deployment}/{instance}')
```

The background for me is the following, we have currently 4 environments and 10 instances (e.g. countries) which results in a huge load of files in one directory and I would like to organize it in a folder structure and hide a bit the complexity :smile:.

Currently it looks a bit like this

```
default.js
default-DE.js
default-FR.js
...
staging.js
staging-DE.js
staging-...
live.js
live-...
```

And 2 more. Maybe I could drop one of the deployments and use more default, but still our config differs by country and having 40 files in one folder seems a bit too much. The order here is quite clear, we could use the same logic as before.

**Comment by markstos** [comment]
We already try to load at least 16 different files, so we aren't keen to add more complexity on this side.

The file system can easily cope with up to 10,000 files in a single directory, so performance is not a concern.

**Comment by evangelion1204** [comment]
> The file system can easily cope with up to 10,000 files in a single directory, so performance is not a concern.

Sure, the file system is no problem, but the human eye and the way you maintain data.

> We already try to load at least 16 different files, so we aren't keen to add more complexity on this side.

Sure this adds again some complexity, but unit tests will take care that everything is working fine. I would agree that it might break users which somehow already use a similar approach like symlinks or requires.

**Comment by markstos** [comment]
I would be more inclined to consider to the change if there was more demand for it. We have over 200,000 downloads per month, the project has been available since 2011, and so far this is this the first and only time this feature request has come up. 

Another alternative is use a build tool to copy or move the configuration files around so that you can store them where you like, but the config system can also find them where expected at run time.

**Comment by evangelion1204** [comment]
> Another alternative is use a build tool to copy or move the configuration files around so that you can store them where you like, but the config system can also find them where expected at run time.

We already have a build tool solution and want to get away from it :smile: no more magic happening, just starting the application with env=live or env=release without any more steps. 

In any case I made a PR earlier, which is still WIP but working in the end. Maybe others will vote for this feature and then you might consider this feature.

**Comment by markstos** [comment]
I appreciate that you contributed some code as well as the request. Yes, if there's more interest in the addition we'll reconsider.

**Comment by sh-guti** [comment]
Hi,
+1 for this.

In our microservice architecture we leverage 10 services, each of them has its own config variables and they've also sharing variables. So for 3 deployments we have ~30 config files.
It would be great to split them to different folders (based on deployment for example)

Anyway, thanks for great library.

**Comment by markstos** [comment]
I have about 160 files in my `config` directory. Though I what I find is the most common way I want to group files together is by customer. Each customer likely has a "default" file a "develop" file and a "production" file.

If I replaced the "dashes" with directories, I would have 3 different folders to look in to see all the files for a  particular customer.

On the other hand, if I made a directory for every customer, I would have something like 40 directories each with old 3 files in it.

So far it hasn't been a problem to have a single config directory. If I want to look at all the files for a customer, I can do ls `./config/*customername*`.
